napari/_qt/qt_viewer.py

```diff
@@ -1050,7 +1053,121 @@ def dropEvent(self, event):
             else:
                 filenames.append(url.toString())
 
-        self.viewer.open(filenames, stack=bool(shift_down))
+        # if trying to open as a stack, just do the old behaviour
+        if shift_down:
+            self.viewer.open(filenames, stack=bool(shift_down))
+            return
+
+        for filename in filenames:
+            _, extension = os.path.splitext(filename)
+            reader_associations = get_settings().plugins.extension2reader
+
+            # get plugin choice from user
+            # choice is None if file was opened or user chose cancel
+            choice = self._get_reader_choice_for_file(
+                filename, extension, reader_associations
+            )
+            if choice:
+                display_name, plugin_name, persist_choice = choice
+                self.viewer.open(
+                    filename,
+                    plugin=plugin_name,
+                )
+                # do we have settings to save?
+                if persist_choice:
+                    # need explicit reassignment for persistence
+                    get_settings().plugins.extension2reader = {
+                        **reader_associations,
+                        extension: display_name,
+                    }
+
+    def _get_reader_choice_for_file(
```

Comment:
We went over this together and I recall you wanted to refactor/test this function. Just wondering what the status was there. 

I think there are some things you might clean up once you start looking at the typing, but afact the logic here is ok.

Reply:
Yes I still need to do this and it's on my list for tomorrow morning! I'll ping you when it's updated